hack/generator/pkg/astmodel/package_import_set.go

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package astmodel
+
+import "sort"
+
+// PackageImportSet represents a set of distinct PackageImport references
+type PackageImportSet struct {
+	imports map[PackageImport]struct{}
+}
+
+// EmptyPackageImportSet creates a new empty set of PackageImport references
+func EmptyPackageImportSet() *PackageImportSet {
+	return &PackageImportSet{
+		imports: make(map[PackageImport]struct{}),
+	}
+}
+
+// AddImport ensures the set includes the specified import
+// Adding an import already present in the set is fine.
+func (set *PackageImportSet) AddImport(packageImport PackageImport) {
+	set.imports[packageImport] = struct{}{}
+}
+
+// AddReference ensures this set includes an import of the specified reference
+// Adding a reference already in the set is fine.
+func (set *PackageImportSet) AddReference(ref PackageReference) {
+	set.AddImport(NewPackageImport(ref))
+}
+
+// Merge ensures that all imports specified in other are included
+func (set *PackageImportSet) Merge(other *PackageImportSet) {
+	for i := range other.imports {
+		set.AddImport(i)
+	}
+}
+
+// Remove ensures the specified item is not present
+// Removing an item not in the set is not an error.
+func (set *PackageImportSet) Remove(packageImport PackageImport) {
+	delete(set.imports, packageImport)
+}
+
+// Contains allows checking to see if an import is included
+func (set *PackageImportSet) ContainsImport(packageImport PackageImport) bool {
+	_, found := set.imports[packageImport]
+	return found
+}
+
+// ImportFor looks up a package reference and returns its import, if any
+func (set *PackageImportSet) ImportFor(ref PackageReference) (PackageImport, bool) {
+	for imp := range set.imports {
```

Comment:
minor: The previous incarnations of code that used this did a lookup in a `map[PackageReference]PackageImport`.

I think this is fine (overall numbers are small), but we are losing a bit of efficiency here and I figured I would call it out in case others objected.

Reply:
I've changed the internal representation of `PackageImportSet` to `map[PackageReference]PackageImport` to allow for this quick lookup.